### PR TITLE
Curvilinear Mesh Interface

### PR DESCRIPTION
**Pull request by sslattery**
This PR adds an interface for general curvilinear mesh in which a mapping function is defined to map from the global physical frame to a global logical frame. This interface is represented by `CurvilinearMeshMapping` which describes all of the functionality needed to compute reference and logical mappings as well as the transformation metrics (e.g. Jacobians). A `CurvilinearMesh` is then built by composing a `Cajita::UniformMesh` in the logical frame and providing access to the mapping function on the device. A default implementation of some mapping functions are provided - particularly for the reference frame mapping in which a Newton method is used.

Two implementations of the mapping interface are provided and more can be added in the future. A uniform Cartesian mapping is given for performance while a bilinear mesh mapping is given to enable general grids and R-adaptivity. In the future we could add things like cylindrical or spherical coordinates. The bilinear mesh mapping relies on knowing the node coordinates as a separate field and has a generator interface which allows one to have different ways to create the coordinates. A default implementation of this is initialization to a uniform grid. In the future we could load a mesh externally.

- The unit tests check the mapping functions for each of the provided mappings. 
- I also updated a few unit tests to handle the recent Cajita changes for the boundary halo.
- The cajita boundary halo actually lets us get rid of the weird internal padding we were doing on non-periodic boundaries so now the input halo width is just the biggest halo that you need. This could be for particle or grid operations. Note that halo widths for the communication options can always be set to less-than-or-equal-to the allocated grid halo width.
- Batched linear algebra was updated so the determinant can be reused for matrix inversions.

Note: `UniformMesh` and `AdaptiveMesh` will eventually be deprecated/removed in favor of the curvilinear mesh. `CurvilinearMesh` will then become the main single block dense mesh class.

### CODE REVIEW

File: unit_test/tstUniformCartesianMeshMapping.hpp [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+    // Global parameters.
+    Kokkos::Array<double, 6> global_box = { -10.0, -10.0, -10.0,
+                                            10.0,  10.0,  10.0 };
+    double cell_size = 0.5;
+    int num_cell = 20.0 / cell_size;
+    int halo_width = 1;
+    Kokkos::Array<bool, 3> periodic = { true, false, true };
+
+    // Partition only in the x direction.
+    std::array<int, 3> ranks_per_dim = { 1, 1, 1 };
```
Comment (kwitaechong) [inline unit_test/tstUniformCartesianMeshMapping.hpp:39]: if partitioned only in x-dir, then ranks_per_dim[0] > 1?

File: src/Picasso_UniformCartesianMeshMapping.hpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                        const ReferenceCoords& reference_coords,
+                        PhysicalCoords& physical_coords )
+    {
+        for ( std::size_t d = 0; d < num_space_dim; ++d )
+            physical_coords( d ) = mapping._global_bounding_box[d] +
+                                   reference_coords( d ) * mapping._cell_size;
+    }
+
+    // Given coordinates in the reference frame compute the grid
+    // transformation metrics. This is the of jacobian of the forward mapping,
```
Comment (kwitaechong) [inline src/Picasso_UniformCartesianMeshMapping.hpp:90]: is the jacobian of the forward mapping

File: unit_test/tstUniformCartesianMeshMapping.hpp [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+        EXPECT_EQ( periodic[d], mapping._periodic[d] );
+    }
+
+    // Check grid.
+    const auto& global_grid = manager->mesh().localGrid()->globalGrid();
+    const auto& global_mesh = global_grid.globalMesh();
+
+    EXPECT_EQ( global_mesh.lowCorner( 0 ), 0.0 );
+    EXPECT_EQ( global_mesh.lowCorner( 1 ), 0.0 );
+    EXPECT_EQ( global_mesh.lowCorner( 2 ), 0.0 );
```
Comment (kwitaechong) [inline unit_test/tstUniformCartesianMeshMapping.hpp:67]: -10?

File: unit_test/tstUniformCartesianMeshMapping.hpp [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+
+    EXPECT_EQ( global_mesh.highCorner( 0 ), num_cell );
+    EXPECT_EQ( global_mesh.highCorner( 1 ), num_cell );
+    EXPECT_EQ( global_mesh.highCorner( 2 ), num_cell );
```
Comment (kwitaechong) [inline unit_test/tstUniformCartesianMeshMapping.hpp:71]: 10?

File: unit_test/tstUniformCartesianMeshMapping.hpp [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        "check_mapping", TEST_EXECSPACE{}, ghosted_cells,
+        KOKKOS_LAMBDA( const int i, const int j, const int k ) {
+            // Map to physical frame.
+            LinearAlgebra::Vector<double, 3> cell_coords;
+            int ijk[3] = { i, j, k };
+            local_mesh.coordinates( Cajita::Cell{}, ijk, cell_coords.data() );
+            LinearAlgebra::VectorView<double, 3> phys_coords(
+                &cell_forward_map( i, j, k, 0 ), cell_forward_map.stride( 3 ) );
+            CurvilinearMeshMapping<mapping_type>::mapToPhysicalFrame(
+                mapping, cell_coords, phys_coords );
```
Comment (kwitaechong) [inline unit_test/tstUniformCartesianMeshMapping.hpp:113]: isn't the cell_coords physical coordinate of local_mesh?

File: unit_test/tstUniformCartesianMeshMapping.hpp (hunk 1 above)
Comment (sslattery) [reply unit_test/tstUniformCartesianMeshMapping.hpp:39]: can be `>=`

File: unit_test/tstUniformCartesianMeshMapping.hpp (hunk 3 above)
Comment (sslattery) [reply unit_test/tstUniformCartesianMeshMapping.hpp:67]: The Cajita mesh is in the logical frame so it spans `(0.0,NumCell)` in each dimension

File: unit_test/tstUniformCartesianMeshMapping.hpp (hunk 4 above)
Comment (sslattery) [reply unit_test/tstUniformCartesianMeshMapping.hpp:71]: The Cajita mesh is in the logical frame so it spans `(0.0,NumCell)` in each dimension

File: unit_test/tstUniformCartesianMeshMapping.hpp (hunk 5 above)
Comment (kwitaechong) [inline unit_test/tstUniformCartesianMeshMapping.hpp:113]: does cell_coords already have physical coordinate by local_mesh.coordinates( Cajita::Cell{}, ijk, cell_coords.data() )?

File: unit_test/tstUniformCartesianMeshMapping.hpp [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+
+            // Map to reference frame.
+            LinearAlgebra::VectorView<double, 3> ref_coords(
+                &cell_reverse_map( i, j, k, 0 ), cell_reverse_map.stride( 3 ) );
+            ref_coords = { cell_coords( Dim::I ) - 0.1,
```
Comment (kwitaechong) [inline unit_test/tstUniformCartesianMeshMapping.hpp:118]: do we need this here?

File: src/Picasso_BilinearMeshMapping.hpp [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+
+        physical_coords = 0.0;
+
+        for ( int ni = 0; ni < 2; ++ni )
+            for ( int nj = 0; nj < 2; ++nj )
+                for ( int nk = 0; nk < 2; ++nk )
+                    for ( int d = 0; d < 3; ++d )
+                        physical_coords( d ) +=
+                            w[Dim::I][ni] * w[Dim::J][nj] * w[Dim::K][nk] *
+                            mapping._local_node_coords( i + ni, j + nj, k + nk,
```
Comment (kwitaechong) [inline src/Picasso_BilinearMeshMapping.hpp:130]: I think local_node_coords is in physical reference since x = sum g_i * N_i

File: src/Picasso_BilinearMeshMapping.hpp [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+    }
+
+    // Get the peridicity of each logical dimension of the mesh.
+    static Kokkos::Array<bool, num_space_dim>
+    periodic( const Generator& generator )
+    {
+        return generator.periodic;
+    }
+
+    // Given the local grid compute populate the local node coordinates. 3D
```
Comment (kwitaechong) [inline src/Picasso_BilinearMeshMapping.hpp:448]: compute the local node coordinate

File: src/Picasso_BilinearMeshMapping.hpp [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+    // Specialization.
+    template <class NodeCoordinateArray, std::size_t NSD = NumSpaceDim>
+    static std::enable_if_t<3 == NSD, void>
+    createLocalNodeCoordinates( const Generator& generator,
+                                const NodeCoordinateArray& coords )
```
Comment (kwitaechong) [inline src/Picasso_BilinearMeshMapping.hpp:453]: when view of coords changes its value, is it still `const`?

File: unit_test/tstUniformCartesianMeshMapping.hpp (hunk 6 above)
Comment (kwitaechong) [inline unit_test/tstUniformCartesianMeshMapping.hpp:118]: do we need this here?

File: unit_test/tstUniformCartesianMeshMapping.hpp [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+                EXPECT_FLOAT_EQ( cell_size * ( global_offset_j + j + 0.5 ) +
+                                     global_box[1],
+                                 forward_map_h( i, j, k, Dim::J ) );
+                EXPECT_FLOAT_EQ( cell_size * ( global_offset_k + k + 0.5 ) +
+                                     global_box[2],
+                                 forward_map_h( i, j, k, Dim::K ) );
+
+                EXPECT_TRUE( map_success_h( i, j, k ) );
+
+                EXPECT_FLOAT_EQ( global_offset_i + i + 0.5,
```
Comment (kwitaechong) [inline unit_test/tstUniformCartesianMeshMapping.hpp:174]: and this equals to cell_coods(0)?

File: unit_test/tstAPIC.hpp [hunk 11]
```diff
@@ -583,10 +583,13 @@ void collocatedTest()
         "fill_grid_vector", TEST_EXECSPACE(),
         grid_vector->layout()->indexSpace( Cajita::Own(), Cajita::Local() ),
         KOKKOS_LAMBDA( const int i, const int j, const int k, const int d ) {
-            gv_view( i, j, k, d ) = 0.0000000001 * ( d + 1 ) * pow( i, 5 ) -
-                                    0.0000000012 * pow( ( d + 1 ) + j * i, 3 ) +
-                                    0.0000000001 * pow( i * j * k, 2 ) +
-                                    pow( ( d + 1 ), 2 );
+            int ic = i - 2;
+            int jc = j - 2;
+            int kc = k - 2;
```
Comment (kwitaechong) [inline unit_test/tstAPIC.hpp:588]: after

File: unit_test/tstAPIC.hpp [hunk 12]
```diff
@@ -583,10 +583,13 @@ void collocatedTest()
...
+            gv_view( i, j, k, d ) =
+                0.0000000001 * ( d + 1 ) * pow( ic, 5 ) -
+                0.0000000012 * pow( ( d + 1 ) + jc * ic, 3 ) +
+                0.0000000001 * pow( ic * jc * kc, 2 ) + pow( ( d + 1 ), 2 );
```
Comment (kwitaechong) [inline unit_test/tstAPIC.hpp:592]: any reason of this changes? To consider the halo when it is non-periodic boundary condition? And Mathematica still remains same?

**Review by kwitaechong — APPROVED** [review]

File: src/Picasso_UniformCartesianMeshMapping.hpp (hunk 2 above)
Comment (sslattery) [reply src/Picasso_UniformCartesianMeshMapping.hpp:90]: fixed

File: src/Picasso_BilinearMeshMapping.hpp (hunk 8 above)
Comment (sslattery) [reply src/Picasso_BilinearMeshMapping.hpp:448]: fixed

File: unit_test/tstUniformCartesianMeshMapping.hpp (hunk 6 above)
Comment (sslattery) [reply unit_test/tstUniformCartesianMeshMapping.hpp:118]: Good question. The container (i.e. the `Kokkos::View`) is const but the data that it points to is not. Note the the definition of all `operator()` in a `Kokkos::View` is actually `const` to allow for inlining.

File: unit_test/tstUniformCartesianMeshMapping.hpp (hunk 10 above)
Comment (sslattery) [reply unit_test/tstUniformCartesianMeshMapping.hpp:174]: correct

File: unit_test/tstAPIC.hpp (hunk 11 above)
Comment (sslattery) [reply unit_test/tstAPIC.hpp:588]: This change was to deal with the new boundary halo in Cajita. Hence the local grid was offset by the halo size, even if there is no periodic boundary.

File: unit_test/tstAPIC.hpp (hunk 12 above)
Comment (sslattery) [reply unit_test/tstAPIC.hpp:592]: Exactly. As I commented above we now have to account for the new Cajita boundary halo region for non-periodic boundaries.
